posts/2019-07-04-Rust-1.36.0.md

```diff
@@ -0,0 +1,164 @@
+
+[alloc-crate]: https://doc.rust-lang.org/alloc/index.html
+
+Before 1.36.0, the standard library consisted of the crates `std`, `core`, and `proc_macro`. The `core` crate provided core functionality such as `Iterator` and `Copy` and could be used in `#![no_std]` environments since it did not impose any requirements. Meanwhile, the `std` crate provided types like `Box<T>` and OS functionality but required a global allocator and other OS capabilities in return.
+
+Starting with Rust 1.36.0, the parts of `std` that depend on a global allocator, e.g. `Vec<T>`, are now defined in the `alloc` crate. The `std` crate then re-exports these parts. As a result, `#![no_std]` applications defining an allocator can use libraries that depend on the `alloc` crate.  While `#![no_std]` applications still require nightly, our hope is that a `#![no_std]` compatible ecosystem of libraries can be built up meanwhile.
+
+If you are the maintainer of a library that only relies on some allocation primitives to function, consider making your library `#[no_std]` compatible by using the following at the top of your `lib.rs` file:
+
+```rust
+#![no_std]
+
+extern crate alloc;
+
+use alloc::vec::Vec;
+```
+
+### [`MaybeUninit<T>`] instead of [`mem::uninitialized`]
+
+[`MaybeUninit<T>`]: https://doc.rust-lang.org/std/mem/union.MaybeUninit.html
+[`mem::uninitialized`]: https://doc.rust-lang.org/std/mem/fn.uninitialized.html
+[gankro-blog]: https://gankro.github.io/blah/initialize-me-maybe/
+[pr-60445]: https://github.com/rust-lang/rust/pull/60445
+
+In previous releases of Rust, the [`mem::uninitialized`] has allowed you bypass Rust's initialization checks by pretending that you've initialized a value at type `T` without doing anything. One of the main uses of this function has been to lazily allocate arrays.
+
+However, [`mem::uninitialized`] is an incredibly dangerous operation that essentially cannot be used correctly as the Rust compiler assumes that values are properly initialized. For example, calling `mem::uninitialized::<bool>()` causes *instantaneous __undefined behavior__* as the random bits in memory could be something other than `0` and `1`.
+
+To remedy this situation, in Rust 1.36.0, the type [`MaybeUninit<T>`] has been [stabilized][pr-60445]. The Rust compiler will understand that it should not assume that a [`MaybeUninit<T>`] is a properly initialized `T`. Therefore, you can do gradual initialization more safely and eventually use `.assume_init()` once you are certain that `maybe_t: MaybeUninit<T>` contains an initialized `T`.
+
+As [`MaybeUninit<T>`] is the safer alternative, starting with Rust 1.38, the function [`mem::uninitialized`] will be deprecated.
+
+To read more about uninitialized memory, [`mem::uninitialized`], and [`MaybeUninit<T>`], read [Alexis Beingessner's blog post][gankro-blog] about the subject. The standard library also contains extensive documentation about [`MaybeUninit<T>`].
+
+### NLL on Rust 2015
```

Comment:
I think "for" is more appropriate than "on"; I'll go ahead and submit suggestions for this change in verbiage throughout the post, so be sure to either accept or reject all of them together.

```suggestion
### NLL for Rust 2015
```